cloud-storage/cloud_storage.c

```diff
@@ -0,0 +1,947 @@
+         if (operation->parameters.upload.local_file)
+         {
+            free(operation->parameters.upload.local_file);
+         }
+         break;
+      case DOWNLOAD_FILE:
+         if (operation->parameters.download.remote_file)
+         {
+            cloud_storage_item_free(operation->parameters.download.remote_file);
+         }
+         break;
+      default:
+         break;
+   }
+
+   free(operation);
+}
+
+static void _free_operation_fn(void *operation)
+{
+   _free_operation((struct _operation_t *) operation);
+}
+
+static void _operation_thread_loop(void *user_data)
+{
+   while (!_shutdown)
+   {
+      struct _operation_t *operation;
+
+      /* In the critical section:
+       * - Go back to waiting if the opertion queue is empty
+       * - Break out of the loop and shutdown the operation thread if
+       *   _shutdown is true
+       * - Pop an operation off of the queue and release the lock
+       */
+      slock_lock(_operation_queue_mutex);
+      if (generic_queue_length(_operation_queue) == 0)
+      {
+         scond_wait(_operation_condition, _operation_queue_mutex);
+      }
+
+      if (_shutdown)
+      {
+         slock_unlock(_operation_queue_mutex);
+         continue;
+      }
+
+      operation = (struct _operation_t *)generic_queue_pop(_operation_queue);
+      slock_unlock(_operation_queue_mutex);
+
+      if (operation)
+      {
+         switch (operation->operation_type)
+         {
+            case SYNC_FILES:
+               _sync_files_execute(operation->parameters.sync.folder_type);
+               break;
+            case UPLOAD_FILE:
+               _upload_file_execute(operation->parameters.upload.folder_type, operation->parameters.upload.local_file);
+               break;
+            case DOWNLOAD_FILE:
+               break;
+         }
+
+         _free_operation(operation);
+      }
+   }
+
+   /* Broke out of the operation loop, will shutdown the operation thread */
+   if (_operation_queue)
+   {
+      generic_queue_free(_operation_queue, &_free_operation_fn);
+   }
+}
+
+void cloud_storage_init(void)
+{
+   struct _operation_queue_item *sync_item;
+
+   if (!_shutdown)
+   {
+      return;
+   }
+
+   _provider = cloud_storage_local_folder_create();
+
+   _operation_queue = generic_queue_new();
+   _operation_queue_mutex = slock_new();
+   _operation_condition = scond_new();
+
+   _shutdown = false;
+   _operation_thread = sthread_create(_operation_thread_loop, NULL);
+
+   if (_provider->ready_for_request())
+   {
+      cloud_storage_sync_files();
+   }
+}
+
+void cloud_storage_shutdown(void)
+{
+   if (_shutdown)
+   {
+      return;
+   }
+
+   /* Signal the operation thread to shutdown in a critical section */
+   slock_lock(_operation_queue_mutex);
+   _shutdown = true;
+   scond_signal(_operation_condition);
+   slock_unlock(_operation_queue_mutex);
+
+   sthread_join(_operation_thread);
+}
+
+char *cloud_storage_get_md5_hash(char *absolute_filename)
```

Comment:
Could you please move this function to the `libretro-common/hash/lrc_hash.c` file? Or somewhere in common. There's already a function that does the same for SHA1 hashes (returns a hash for a given file path).

Reply:
Moved.